### ISSUE TITLE
A/B testing on scan pause command

### ISSUE THREAD

**Issue by ericdill** [NSLS-II/bluesky#152]
@tacaswell and I are having a disagreement on what hitting `ctrl+c` should do.

@tacaswell wants this behavior:
```
...scan is running
[user input] ctrl+c captured
...behind the scenes there is a 0.5 second timeout that is waiting for another 
   ctrl+c. If this other ctrl+c is captured, the scan will end with an Abort state
...if no additional ctrl+c is received within 0.5 seconds, the user is dropped 
   back to the IPython REPL
```

@ericdill wants this behavior:
```
...scan is running
[user input] ctrl+c captured

You pressed ctrl+c. Do you want to:
(r) resume the scan                                                     
(a) abort the scan (bad exit state)
(s) stop the scan (good exit state)
(p) pause the scan (and return to the REPL)

Note that an additional ctrl+c will (a)bort the scan.
input --> 

...indefinitely wait for user input.
```

It seems like the right approach here is to do some A/B testing!  Build both and see which one users like!

cc @danielballan

**Comment by danielballan** [comment]
Sure, as long as we pick a winner in the near future.

I implemented something like the second way on a branch awhile back, except
that the pause happens immediately and the user only gets to choose what
happens next.

I came around to the first way yesterday in conversation with Claudio,
Stuart, and Andy.

I'll show them both and see what they like.

**Comment by ericdill** [comment]
My real concern with the first option above is that the user will be returned to the REPL with no clear indication of what to do next.  That concern is the **entire** motivation with the (perhaps unecessarily) complicated input prompt I proposed.

How about this. I am fully on-board with @tacaswell 's option above (which might actually be @danielballan 's suggestions, I'm not sure how the credit falls out here!), but print out a few lines that is (roughly) like this and then return the user to the REPL with the run engine in a `paused` state. Unless two ctrl+c's are intercepted which should (abort or stop?) the scan.
```
Your RunEngine is now in a paused state. These are your options for changing 
the state of the RunEngine:

resume()  --> will resume the scan
 abort()  --> will kill the scan with an 'aborted' state to indicate 
              the scan was interrupted
  stop()  --> will kill the scan with a 'finished' state to indicate 
              the scan stopped normally
```

**Comment by danielballan** [comment]
I think you are onto something. Let me spell it out in tedious detail to
make sure we are on the same page.

1. The RunEngine catches a Ctrl+C signal.
2. The RunEngine starts listening for a second Ctrl+C signal.
3. The RunEngine requests a pause.
4. If a pause is not possible, the RunEngine aborts, becomes idle, and
returns the REPL. A message is printed, the same as it is now, and we are
done.
5. If a pause is possible, the RunEngine enters a paused state and
withholds the REPL for half a second, enough time for a second Ctrl+C to
register. If a Ctrl+C is caught, the RunEngine aborts, as in (4).
6. Otherwise, the RunEngine prints a message like the one you suggested
above.

I like the half-second pause because it removes any doubt about whether the
RunEngine is blocking Ctrl+C or not: if the user has the REPL, the user
owns Ctrl+C. If not, the RunEngine owns it.

Dan

**Comment by ericdill** [comment]
Other than your step two being out of order (step two is really part of step five), yes we are in agreement. Thanks for spelling this out explicitly!

**Comment by danielballan** [comment]
You might be right.

My thought was this: Assuming that aborting takes some time though, and
assuming that the user might mash Ctrl+C twice out of habit, it might
correct to catch and ignore a second Ctrl+C, if it occurs before we are
ready to return the REPL.

I'm not sure that makes sense, though. We'll see how the implementation
shakes out.

Note that, no matter what, if the user hits Ctrl+C *three *times, we won't
be getting in the way at that point. So it's still possible to kill
stubborn scans by mashing Ctrl+C frantically -- an important feature!

**Comment by danielballan** [comment]
Closed by #154
